File: ts/128/19.ts
import { ListNode } from "../class"

// 双指针
function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
  const dummy = new ListNode(-1, head)
  let fast = head, slow = dummy, i = 0
  while (i < n) {
    fast = fast.next
    i += 1
  }
  while (fast) {
    fast = fast.next
    slow = slow.next
  }
  slow.next = slow.next.next
  return dummy.next
};

File: c/128/19.c
#include <stdio.h>
#include <stdlib.h>
#include "../structs.h"

struct ListNode* removeNthFromEnd(struct ListNode* head, int n){
  struct ListNode* dummy = malloc(sizeof(struct ListNode));
  dummy->val = -1, dummy->next = head;
  struct ListNode* fast = head;
  struct ListNode* slow = dummy;
  for (int i = 0; i < n; i++) {
    fast = fast->next;
  }

  while (fast != NULL) {
    fast = fast->next;
    slow = slow->next;
  }

  slow->next = slow->next->next;
  return dummy->next;
}